ckanext/canada/helpers.py

```diff
@@ -719,6 +728,86 @@ def flash_success(message, allow_html=True):
               allow_html=allow_html)
 
 
+def get_loader_status_badge(resource):
+    # type: (dict) -> str
+    """
+    Displays a custom badge for the status of Xloader and DataStore
+    for the specified resource.
+    """
+
+    if not XLoaderFormats:
+        return ''
+
+    if not resource.get('url_type') == 'upload' or \
+    not XLoaderFormats.is_it_an_xloader_format(resource.get('format')):
+        # we only want to show badges for uploads of supported xloader formats
+        return ''
+
+    is_datastore_active = resource.get('datastore_active', False)
+
+    try:
+        xloader_job = t.get_action("xloader_status")(None, {"resource_id": resource.get('id')})
+    except (t.ObjectNotFound, t.NotAuthorized):
+        xloader_job = {}
+
+    if xloader_job.get('status') == 'complete':
+        # the xloader task is complete, show datastore active or inactive.
+        # xloader will delete the datastore table at the beggining of the job run.
+        # so this will only be true if the job is fully finished.
+        status = 'active' if is_datastore_active else 'inactive'
+    elif xloader_job.get('status') in ['pending', 'running', 'running_but_viewable', 'error']:
+        # the job is running or pending or errored
+        # show the xloader status
+        status = xloader_job.get('status')
+        if status == 'running_but_viewable':
+            # treat running_but_viewable the same as running
+            status = 'running'
+    else:
+        # we do not know what the status is
+        status = 'unknown'
+
+    messages = {
+        'pending': _('Data awaiting load to DataStore'),
+        'running': _('Loading data into DataStore'),
+        'complete': _('Data loaded into DataStore'),
+        'error': _('Failed to load data into DataStore'),
+        'active': _('Data available in DataStore'),
+        'inactive': _('Resource not active in DataStore'),
+        'unknown': _('DataStore status unknown'),
+    }
+
+    pusher_url = t.h.url_for('xloader.resource_data',
+                             id=resource.get('package_id'),
+                             resource_id=resource.get('id'))
+
+    badge_url = t.h.url_for_static('/static/img/badges/{lang}/datastore-{status}.svg'.format(lang=t.h.lang(), status=status))
+
+    # in queue, try to get the number in the queue
+    # NOTE: the Validation extension does not use the task_status table,
+    # and we will never know the job id of a Validation job.
+    # FIXME: is this worth it if we cannot display in the Xloader page?
+    # is it worth it if we cannot display for the Validation extension?
+    queue_number = None
+    if status == 'pending':
+        queue_number = 1
+        jobs_in_queue = t.get_action('job_list')({"ignore_auth": True}, {})
+        for job in jobs_in_queue:
+            if job.get('id') == xloader_job.get('job_id'):
+                break
+            queue_number += 1
```

Comment:
Ah yes...well funny thing about that. I came across that issue when our xloader queue got backed up to 4k jobs. The `job_list` action method returns a dictization of every single job in the queue. Which I suppose might be an issue/improvement in Core to make.

So I tested it by loading up 800 jobs in the queue locally and it really broke the Job Queue admin page I was working on. So that is why in the Job Queue view, I do `MAX_JOB_QUEUE_LIST` of 25 to output on the page.

In the RQ class from redis lib, there is a `job_ids` class property, CKAN Core just uses the `jobs` prop and dictizes each job object. But another action method in ckan core to just return the job_ids list could be useful too.

But for now, and for this, I did put in the `FIXME` to discuss if getting the number in queue was even worth it. Especially as we cannot know a Validation job's id without being inside of the loop of job objects. So it does not really seem worth it, especially in the badges which show up on critical UX pages.

So removing the queue count I think for now.